libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp

```diff
@@ -1021,6 +1024,107 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
     storedExtNavVel.push(extNavVelNew);
 }
 
+/*
+  update the GPS selection
+ */
+void NavEKF3_core::update_gps_selection(void)
+{
+    auto &gps = AP::gps();
+
+    // in normal operation use the primary GPS
+    selected_gps = gps.primary_sensor();
+    preferred_gps = selected_gps;
+
+    if (frontend->_affinity & EKF_AFFINITY_GPS) {
+        if (core_index < gps.num_sensors() ) {
+            // always prefer our core_index, unless we don't have that
+            // many GPS sensors available
+            preferred_gps = core_index;
+        }
+        if (gps.status(preferred_gps) >= AP_GPS::GPS_OK_FIX_3D) {
+            // select our preferred_gps if it has a 3D fix, otherwise
+            // use the primary GPS
+            selected_gps = preferred_gps;
+        }
+    }
+}
+
+/*
+  update the mag selection
+ */
+void NavEKF3_core::update_mag_selection(void)
+{
+    const auto *compass = _ahrs->get_compass();
+    if (compass == nullptr) {
+        return;
+    }
+
+    // in normal operation use the primary compass
```

Comment:
this breaks the existing compass switching when affinity is disable for compasses (note that my PR broke that too, I've now fixed it).
When affinity is disabled we want to use the existing compass switching logic which is based on innovations, so just delete this line and leave magSelectIndex as it is

Reply:
fixed here now